py/map.c

```diff
@@ -172,6 +203,7 @@ mp_map_elem_t *mp_map_lookup(mp_map_t *map, mp_obj_t index, mp_map_lookup_kind_t
                     elem->value = value;
                 }
                 #endif
+                MAP_CACHE_SET(index, elem - map->table);
```

Comment:
Maybe this should be wrapped in `else { ... }`, because if it was a `REMOVE_IF_FOUND` then the cache should not be updated (with a now-removed element).

Reply:
This doesn't seem to help... I guess the relatively uncommon removes are outweighed by the extra branch on the common path.

```
$ ./run-perfbench.py -s ~/3.cache-map-add-pybv11 ~/3.cache-map-add-else-pybv11 
diff of scores (higher is better)
N=100 M=100              /home/jimmo/3.cache-map-add-pybv11 -> /home/jimmo/3.cache-map-add-else-pybv11         diff      diff% (error%)
bm_asyncio.py                113.74 ->     113.72 :      -0.02 =  -0.018% (+/-0.01%)
bm_chaos.py                  343.50 ->     336.44 :      -7.06 =  -2.055% (+/-0.00%)
bm_fannkuch.py                78.62 ->      78.26 :      -0.36 =  -0.458% (+/-0.01%)
bm_fft.py                   2593.13 ->    2552.65 :     -40.48 =  -1.561% (+/-0.00%)
bm_float.py                 5667.22 ->    5569.22 :     -98.00 =  -1.729% (+/-0.00%)
bm_hexiom.py                  42.58 ->      42.75 :      +0.17 =  +0.399% (+/-0.00%)
bm_nqueens.py               4464.29 ->    4455.36 :      -8.93 =  -0.200% (+/-0.00%)
bm_pidigits.py               649.28 ->     650.14 :      +0.86 =  +0.132% (+/-0.37%)
misc_aes.py                  425.96 ->     420.22 :      -5.74 =  -1.348% (+/-0.01%)
misc_mandel.py              3620.72 ->    3562.43 :     -58.29 =  -1.610% (+/-0.00%)
misc_pystone.py             2368.03 ->    2342.79 :     -25.24 =  -1.066% (+/-0.01%)
misc_raytrace.py             354.92 ->     351.44 :      -3.48 =  -0.981% (+/-0.01%)
```